Cognite.Simulator.Utils/FileStorageClient.cs

```diff
@@ -29,38 +33,65 @@ public FileStorageClient(HttpClient client, ILogger<FileStorageClient> logger)
-        public async Task<bool> DownloadFileAsync(Uri uri, string filePath)
+        public async Task<bool> DownloadFileAsync(Uri uri, string filePath, bool failLargeFiles = false)
         {
             try
             {
-                var response = await _client.GetAsync(uri)
+                var response = await _client.GetAsync(uri, HttpCompletionOption.ResponseHeadersRead)
                     .ConfigureAwait(false);
                 if (!response.IsSuccessStatusCode)
                 {
                     _logger.LogWarning("Failed to download the file: {Message}", response.ReasonPhrase);
                     return false;
                 }
 
+                if (response.Content.Headers.ContentLength > MaxFileDownloadSize)
+                {
+                    _logger.LogWarning("File size exceeds the maximum allowed size: {Size} bytes", MaxFileDownloadSize);
+                    return false;
+                }
+
+                if (failLargeFiles && response.Content.Headers.ContentLength > LargeFileSize)
+                {
+                    throw new ConnectorException(
+                        $"The file is larger than {LargeFileSize / 1024 / 1024} MB and needs to be pre-downloaded before any operations can be performed." +
+                        "Connector hasn't been able to pre-download the file yet. " +
+                        "Please wait for the file to be downloaded and try again."
+                    );
+                }
+
                 if (File.Exists(filePath))
                 {
                     File.Delete(filePath);
                 }
-                using (var fs = new FileStream(filePath, FileMode.CreateNew))
+
+                using (var responseStream = await response.Content.ReadAsStreamAsync().ConfigureAwait(false))
+                using (var fileStream = new FileStream(filePath, FileMode.CreateNew, FileAccess.Write, FileShare.None, bufferSize: DefaultBufferSize, useAsync: true))
                 {
-                    await response.Content.CopyToAsync(fs)
-                        .ConfigureAwait(false);
+                    var buffer = new byte[DefaultBufferSize];
+                    int bytesRead;
+
+                    while ((bytesRead = await responseStream.ReadAsync(buffer, 0, buffer.Length).ConfigureAwait(false)) > 0)
+                    {
+                        await fileStream.WriteAsync(buffer, 0, bytesRead).ConfigureAwait(false);
+                    }
                     return true;
                 }
             }
             catch (HttpRequestException e)
             {
                 // File cannot be downloaded, skip for now and try again later
-                _logger.LogError("Failed to download the file: {Message}", e.Message);
-                return false;
+                _logger.LogError("Failed to download the file {uri}: {Message}", uri, e.Message);
+            }
+            catch (IOException e)
```

Comment:
Should we also catch generic Exceptions here?

Reply:
it's caught one level above on the callee in `ModelLibraryBase.cs`